File: src/004. Largest palindrome product/004.lua
local max = 0
for a = 999, 100, -1 do
  for b = 999, 100, -1 do
    local num = a * b
    if num == tonumber(tostring(num):reverse()) then
      if num > max then
        max = num
      end
    end
  end
end
print(max)

File: src/004. Largest palindrome product/004.cpp
#include <iostream>
#include <string>
using std::string;

bool is_palindrome(long long num);

int main()
{
    long long max = 0;
    for (int i = 999; i >= 100; --i)
    {
        for (int j = 999; j >= 100; --j)
        {
            long long product = i * j;
            if (is_palindrome(product))
            {
                if (product > max)
                {
                    max = product;
                }
            }
        }
    }
    std::cout << max << std::endl;
}

bool is_palindrome(long long num)
{
    string original = std::to_string(num);
    string reversed(original.rbegin(), original.rend());
    return original == reversed;
}
